python/sglang/srt/entrypoints/openai/serving_chat.py

```diff
@@ -75,6 +80,23 @@ def _validate_request(self, request: ChatCompletionRequest) -> Optional[str]:
         ):
             return "Tools cannot be empty if tool choice is set to required."
 
+        if request.tool_choice is not None and not isinstance(request.tool_choice, str):
+            if not request.tools:
+                return "Tools cannot be empty if tool choice is set to a specific tool."
+            tool_name = request.tool_choice.function.name
+            tool_exists = any(tool.function.name == tool_name for tool in request.tools)
+            if not tool_exists:
+                return f"Tool '{tool_name}' not found in tools list."
+
+        # Validate tool definitions
```

Comment:
This might reduce the indentation more.

```suggestion
for i, tool in enumerate(request.tools or []):
    if tool.function.parameters is None:
        continue
    try:
        Draft202012Validator.check_schema(tool.function.parameters)
    except SchemaError as e:
        return f"Tool {i} function has invalid 'parameters' schema: {e}"
```

Reply:
Note: we can't commit directly as I didn't wrap all the lines.